idseq_dag/util/server.py

```diff
@@ -173,6 +185,42 @@ def wait_for_server_ip(service_name,
         # if we had to wait here, that counts toward the rate limit delay
         result = wait_for_server_ip_work(service_name, key_path,
                                          remote_username, environment,
-                                         max_concurrent, chunk_id)
+                                         max_concurrent, chunk_id,
+                                         max_interval_between_describe_instances,
+                                         draining_tag)
         return result
 
+
+def build_job_tag(job_tag_prefix, chunk_id):
+    batch_job_id = os.environ.get('AWS_BATCH_JOB_ID', 'local')
+    job_tag_key = f"{job_tag_prefix}{batch_job_id}_chunk{chunk_id}"
+    job_tag_value = int(time.time())
+    return job_tag_key, job_tag_value
+
+
+def create_tag(instance_iD, tag_key, tag_value):
+    command.execute(f"aws ec2 create-tags --resources {instance_iD} --tags Key={tag_key},Value={tag_value}")
+
+def delete_tag(instance_iD, tag_key):
+    command.execute(f"aws ec2 delete-tags --resources {instance_iD} --tags Key={tag_key}")
+
+
+@contextmanager
+def ASGInstance(service, key_path, remote_username, environment, max_concurrent, chunk_id,
```

Comment:
A further refactor/improvement opportunity would be to create an ASG class, per service, to encapsulate all of these various arguments so that you won't have to pass them as positional arguments through the call chain.

That way, in the function that splits into chunks, you will have one statement like this
```
    asg = ASG(service, key_path, remote_username, ...)
```
so all those vars would be captured in there.   In fact, you probably don't even need to pass them all into the constructor -- just put the code that extracts them from the DAG and ENV into the constructor.

Then at point of use, you can refer to them as asg.service, asg.key_path, etc, etc.    And any function that is currently getting  a list of these as its arguments can now get a single argument called asg.

Reply:
Thanks for this suggestion! In order to keep the current change minimal, I will defer this refactor until after the scaling is in order.